File: config/network_robots/config_loader.js
/**
 * Network Robot Config Loader (JavaScript)
 * 병합 로더: _defaults.yaml + overrides/{robot}.yaml → 완전한 config
 *
 * Usage:
 *   import { loadRobotConfig } from './config_loader.js';
 *   const config = await loadRobotConfig('lite6');
 */

// js-yaml 필요: npm install js-yaml
// 또는 브라우저에서 CDN: <script src="https://cdn.jsdelivr.net/npm/js-yaml@4.1.0/dist/js-yaml.min.js"></script>

/**
 * Deep merge override into base
 * @param {Object} base - Base object
 * @param {Object} override - Override object
 * @returns {Object} Merged object
 */
function deepMerge(base, override) {
    const result = JSON.parse(JSON.stringify(base)); // Deep copy

    for (const key in override) {
        if (override.hasOwnProperty(key)) {
            if (result[key] && typeof result[key] === 'object' && typeof override[key] === 'object'
                && !Array.isArray(result[key]) && !Array.isArray(override[key])) {
                result[key] = deepMerge(result[key], override[key]);
            } else {
                result[key] = JSON.parse(JSON.stringify(override[key]));
            }
        }
    }

    return result;
}

/**
 * Load YAML file
 * @param {string} url - URL to YAML file
 * @returns {Promise<Object>} Parsed YAML content
 */
async function loadYaml(url) {
    const response = await fetch(url);
    if (!response.ok) {
        throw new Error(`Failed to load: ${url}`);
    }
    const text = await response.text();

    // Use js-yaml if available, otherwise simple parse
    if (typeof jsyaml !== 'undefined') {
        return jsyaml.load(text);
    } else if (typeof YAML !== 'undefined') {
        return YAML.parse(text);
    } else {
        // Fallback: assume JSON-compatible YAML
        console.warn('js-yaml not found, attempting JSON parse');
        return JSON.parse(text);
    }
}

/**
 * Get base URL for config files
 * @returns {string} Base URL
 */
function getConfigBaseUrl() {
    // Adjust based on your server setup
    return '/config/network_robots';
}

/**
 * Load robot configuration (defaults + override merged)
 * @param {string} robotModel - Robot model name (e.g., "lite6", "xarm5")
 * @param {Object} options - Options
 * @param {boolean} options.convertToRadians - Convert joint angles to radians
 * @param {string} options.baseUrl - Custom base URL
 * @returns {Promise<Object>} Merged config
 */
async function loadRobotConfig(robotModel, options = {}) {
    const { convertToRadians = false, baseUrl = getConfigBaseUrl() } = options;

    // Load defaults
    const defaults = await loadYaml(`${baseUrl}/_defaults.yaml`);

    // Load override
    const override = await loadYaml(`${baseUrl}/overrides/${robotModel}.yaml`);

    // Merge
    let config = deepMerge(defaults, override);

    // Convert to radians if requested
    if (convertToRadians) {
        config = convertJointsToRadians(config);
    }

    return config;
}

/**
 * Convert joint angles from degrees to radians
 * @param {Object} config - Config object
 * @returns {Object} Config with radians
 */
function convertJointsToRadians(config) {
    const result = JSON.parse(JSON.stringify(config));
    const deg2rad = (deg) => deg * Math.PI / 180;

    if (result.joints) {
        if (result.joints.zero) {
            result.joints.zero = result.joints.zero.map(deg2rad);
        }
        if (result.joints.home) {
            result.joints.home = result.joints.home.map(deg2rad);
        }
        if (result.joints.limits) {
            for (const jointName in result.joints.limits) {
                result.joints.limits[jointName] = result.joints.limits[jointName].map(deg2rad);
            }
        }
    }

    return result;
}

/**
 * Load named_frames section only (for FrameManager)
 * @param {string} robotModel - Robot model name
 * @returns {Promise<Object>} named_frames config
 */
async function loadNamedFrames(robotModel) {
    const config = await loadRobotConfig(robotModel, { convertToRadians: true });
    return config.named_frames || {};
}

/**
 * Load admittance section only
 * @param {string} robotModel - Robot model name
 * @returns {Promise<Object>} admittance config
 */
async function loadAdmittanceConfig(robotModel) {
    const config = await loadRobotConfig(robotModel);
    return config.admittance || {};
}

/**
 * Get list of available robots
 * @param {string} baseUrl - Base URL
 * @returns {Promise<string[]>} List of robot model names
 */
async function getAvailableRobots(baseUrl = getConfigBaseUrl()) {
    // This requires server-side support or a manifest file
    // For now, return hardcoded list
    return [
        'lite6', 'xarm5', 'fr3',
        'ur3e', 'ur5e', 'ur16e',
        'iiwa7', 'med7', 'med14',
        'kinova_gen3_7dof'
    ];
}

// Export for ES6 modules
if (typeof module !== 'undefined' && module.exports) {
    module.exports = {
        loadRobotConfig,
        loadNamedFrames,
        loadAdmittanceConfig,
        getAvailableRobots,
        deepMerge,
        convertJointsToRadians
    };
}

// Export for browser global
if (typeof window !== 'undefined') {
    window.ConfigLoader = {
        loadRobotConfig,
        loadNamedFrames,
        loadAdmittanceConfig,
        getAvailableRobots,
        deepMerge,
        convertJointsToRadians
    };
}

File: config/network_robots/config_loader.py
#!/usr/bin/env python3
"""
Network Robot Config Loader
병합 로더: _defaults.yaml + overrides/{robot}.yaml → 완전한 config

Usage:
    from config_loader import load_robot_config
    config = load_robot_config("lite6")
"""

import os
import copy
import math
from pathlib import Path
from typing import Any, Dict, Optional

import yaml


def deep_merge(base: Dict[str, Any], override: Dict[str, Any]) -> Dict[str, Any]:
    """
    Deep merge override into base.
    override의 값이 base를 덮어씀. nested dict는 재귀적으로 병합.
    """
    result = copy.deepcopy(base)

    for key, value in override.items():
        if key in result and isinstance(result[key], dict) and isinstance(value, dict):
            result[key] = deep_merge(result[key], value)
        else:
            result[key] = copy.deepcopy(value)

    return result


def get_config_dir() -> Path:
    """config 디렉토리 경로 반환"""
    return Path(__file__).parent


def load_yaml(file_path: Path) -> Dict[str, Any]:
    """YAML 파일 로드"""
    with open(file_path, 'r', encoding='utf-8') as f:
        return yaml.safe_load(f) or {}


def load_robot_config(robot_model: str, convert_to_radians: bool = False) -> Dict[str, Any]:
    """
    로봇 설정 로드 (_defaults + override 병합)

    Args:
        robot_model: 로봇 모델명 (예: "lite6", "xarm5", "fr3")
        convert_to_radians: True면 joint angles를 radians로 변환

    Returns:
        병합된 완전한 config dict
    """
    config_dir = get_config_dir()

    # Load defaults
    defaults_path = config_dir / "_defaults.yaml"
    if not defaults_path.exists():
        raise FileNotFoundError(f"Defaults file not found: {defaults_path}")
    defaults = load_yaml(defaults_path)

    # Load override
    override_path = config_dir / "overrides" / f"{robot_model}.yaml"
    if not override_path.exists():
        raise FileNotFoundError(f"Override file not found: {override_path}")
    override = load_yaml(override_path)

    # Merge
    config = deep_merge(defaults, override)

    # Convert degrees to radians if requested
    if convert_to_radians:
        config = convert_joints_to_radians(config)

    return config


def convert_joints_to_radians(config: Dict[str, Any]) -> Dict[str, Any]:
    """
    joints.zero, joints.home, joints.limits를 radians로 변환
    """
    result = copy.deepcopy(config)

    if 'joints' in result:
        joints = result['joints']

        # Convert zero
        if 'zero' in joints and joints['zero']:
            joints['zero'] = [math.radians(deg) for deg in joints['zero']]

        # Convert home
        if 'home' in joints and joints['home']:
            joints['home'] = [math.radians(deg) for deg in joints['home']]

        # Convert limits
        if 'limits' in joints:
            for joint_name, limits in joints['limits'].items():
                joints['limits'][joint_name] = [math.radians(deg) for deg in limits]

    return result


def get_available_robots() -> list:
    """사용 가능한 로봇 목록 반환"""
    config_dir = get_config_dir()
    overrides_dir = config_dir / "overrides"

    if not overrides_dir.exists():
        return []

    return [f.stem for f in overrides_dir.glob("*.yaml")]


def load_named_frames(robot_model: str) -> Dict[str, Any]:
    """
    named_frames 섹션만 추출 (FrameManager용)
    """
    config = load_robot_config(robot_model, convert_to_radians=True)
    return config.get('named_frames', {})


def load_admittance_config(robot_model: str) -> Dict[str, Any]:
    """
    admittance 섹션만 추출
    """
    config = load_robot_config(robot_model)
    return config.get('admittance', {})


# CLI test
if __name__ == "__main__":
    import sys
    import json

    if len(sys.argv) < 2:
        print("Usage: python config_loader.py <robot_model> [--radians]")
        print(f"Available robots: {get_available_robots()}")
        sys.exit(1)

    robot_model = sys.argv[1]
    convert_radians = "--radians" in sys.argv

    try:
        config = load_robot_config(robot_model, convert_to_radians=convert_radians)
        print(f"=== {robot_model} config ===")
        print(yaml.dump(config, default_flow_style=False, allow_unicode=True))
    except FileNotFoundError as e:
        print(f"Error: {e}")
        sys.exit(1)
